hazelcast/src/main/java/com/hazelcast/internal/bootstrap/TpcServerBootstrap.java

```diff
@@ -120,39 +127,75 @@ public void start() {
-                    socket.setRegularSchedule(regularSchedule);
-                    socket.sendBufferSize(sendBufferSize);
-                    socket.receiveBufferSize(receiveBufferSize);
-                    socket.tcpNoDelay(tcpNoDelay);
-                    socket.keepAlive(true);
-                    socket.activate(eventloop);
-                });
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
+            NioAsyncServerSocket serverSocket = NioAsyncServerSocket.open(eventloop);
+            serverSockets.add(serverSocket);
+            int receiveBufferSize = clientSocketConfig.getReceiveBufferSize();
+            int sendBufferSize = clientSocketConfig.getSendBufferSize();
+            serverSocket.receiveBufferSize(receiveBufferSize);
+            serverSocket.reuseAddress(true);
+            clientPortIndex = bind(serverSocket, clientAllowedPorts, clientPortIndex);
+            serverSocket.accept(socket -> {
+                socket.readHandler(readHandlerSuppliers.get(eventloop).get());
+                socket.setWriteThrough(writeThrough);
+                socket.setRegularSchedule(regularSchedule);
+                socket.sendBufferSize(sendBufferSize);
+                socket.receiveBufferSize(receiveBufferSize);
+                socket.tcpNoDelay(tcpNoDelay);
+                socket.keepAlive(true);
+                socket.activate(eventloop);
+            });
+        }
+    }
+
+    // public for testing
+    public TPCSocketConfig getClientSocketConfig() {
+        if (nodeEngine.getConfig().getAdvancedNetworkConfig().isEnabled()) {
+            ServerSocketEndpointConfig endpointConfig = (ServerSocketEndpointConfig) nodeEngine
+                    .getConfig()
+                    .getAdvancedNetworkConfig()
+                    .getEndpointConfigs()
+                    .get(EndpointQualifier.CLIENT);
+
+            if (endpointConfig == null) {
+                return null;
             }
+
+            return endpointConfig.getTpcSocketConfig();
         }
+
+        // unified socket
+        return nodeEngine.getConfig().getNetworkConfig().getTpcSocketConfig();
     }
 
-    @SuppressWarnings("checkstyle:MagicNumber")
-    private int toPort(Address address, int socketId) {
-        return (address.getPort() - DEFAULT_HZ_PORT) * 100 + 11000 + socketId % tpcEngine.eventloopCount();
+    private int bind(NioAsyncServerSocket serverSocket, List<Integer> allowedPorts, int index) {
+        try {
+            serverSocket.bind(new InetSocketAddress(thisAddress.getInetAddress(), allowedPorts.get(index)));
+            return index + 1;
+        } catch (BindException e) {
+            return bind(serverSocket, allowedPorts, index + tpcEngine.eventloopCount());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } catch (IndexOutOfBoundsException e) {
+            throw new HazelcastException("Allowed TPC ports weren't enough.");
```

Comment:
IndexOutOfBoundsException should not be abused for normal flow. IndexOutOfBoundsException indicates a bug.

Reply:
https://github.com/hazelcast/hazelcast/pull/23254/commits/393aa24e2f335e7f8fdd0092c7010706b9b3b875